syntax = "proto2";

package com.scalableminds.webknossos.datastore;

import "geometry.proto";
import "MetadataEntry.proto";

message Node {
  required int32 id = 1;
  required Vec3IntProto position = 2;
  required Vec3DoubleProto rotation = 3;
  required float radius = 4;
  required int32 viewport = 5;
  required int32 resolution = 6;
  required int32 bitDepth = 7;
  required bool interpolation = 8;
  required int64 createdTimestamp = 9;
  repeated AdditionalCoordinateProto additionalCoordinates = 10;
}

message Edge {
  required int32 source = 1;
  required int32 target = 2;
}

message Comment {
  required int32 nodeId = 1;
  required string content = 2;
}

message BranchPoint {
  required int32 nodeId = 1;
  required int64 createdTimestamp = 2;
}

enum TreeTypeProto {
  DEFAULT = 0;
  AGGLOMERATE = 1;
}

message Tree {
  required int32 treeId = 1;
  repeated Node nodes = 2;
  repeated Edge edges = 3;
  optional ColorProto color = 4;
  repeated BranchPoint branchPoints = 5;
  repeated Comment comments = 6;
  required string name = 7;
  required int64 createdTimestamp = 8;
  optional int32 groupId = 9;
  optional bool isVisible = 10; // `None` means `true`
  optional TreeTypeProto type = 11;
  optional bool edgesAreVisible = 12; // `None` means `true`
  repeated MetadataEntryProto metadata = 13;
}

message TreeGroup {
  required string name = 1;
  required int32 groupId = 2;
  repeated TreeGroup children = 3;
  optional bool isExpanded = 4;
}

message SkeletonTracing {
  required string datasetId = 1;
  required string datasetName = 2; // used when parsing and handling nmls, not used in tracing store anymore, do not rely on correct values
  repeated Tree trees = 3;
  required int64 createdTimestamp = 4;
  optional BoundingBoxProto boundingBox = 5;
  optional int32 activeNodeId = 6;
  required Vec3IntProto editPosition = 7;
  required Vec3DoubleProto editRotation = 8;
  required double zoomLevel = 9;
  required int64 version = 10;
  optional BoundingBoxProto userBoundingBox = 11;
  repeated TreeGroup treeGroups = 12;
  repeated NamedBoundingBoxProto userBoundingBoxes = 13;
  optional string organizationId = 14; // used when parsing and handling nmls, not used in tracing store anymore, do not rely on correct values
  repeated AdditionalCoordinateProto editPositionAdditionalCoordinates = 21;
  repeated AdditionalAxisProto additionalAxes = 22; // Additional axes for which this tracing is defined
}

message SkeletonTracingOpt {
  optional SkeletonTracing tracing = 1;
}

message SkeletonTracings {
  repeated SkeletonTracingOpt tracings = 1;
}
